interfaces/builtin/pkcs11.go

```diff
@@ -0,0 +1,160 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	apparmor_sandbox "github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const pkcs11Summary = `allows use of pkcs11 framework and access to exposed tokens`
+
+const pkcs11BaseDeclarationSlots = `
+  pkcs11:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+type pkcs11Interface struct{}
+
+// Name of the pkcs11 interface.
+func (iface *pkcs11Interface) Name() string {
+	return "pkcs11"
+}
+
+func (iface *pkcs11Interface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              pkcs11Summary,
+		BaseDeclarationSlots: pkcs11BaseDeclarationSlots,
+	}
+}
+
+func (iface *pkcs11Interface) String() string {
+	return iface.Name()
+}
+
+const pkcs11PermanentSlotSecComp = `
+# Description: Allow operating as an p11-kit server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+
+# Needed by server upon client connect
+accept
+accept4
+`
+
+func (iface *pkcs11Interface) getSocketPath(slot *snap.SlotInfo) (string, error) {
+	socketAttr, isSet := slot.Attrs["pkcs11-socket"]
+	if !isSet {
+		return "", fmt.Errorf(`cannot use pkcs11 slot %s without "pkcs11-socket" attribute`, slot.Name)
+	}
+
+	socketPath, ok := socketAttr.(string)
+	if !ok {
+		return "", fmt.Errorf(`pkcs11: "pkcs11-socket" attribute must be a string, not %v`,
+			slot.Attrs["pkcs11-socket"])
+	}
+
+	if ok := cleanSubPath(socketPath); !ok {
+		return "", fmt.Errorf("slot %q, a unix socket path is not clean", slot.Name)
+	}
+
+	// separate socket name and check socket is in /run/p11-kit
+	if filepath.Dir(socketPath) != "/run/p11-kit" {
+		return "", fmt.Errorf("slot %q, a unix socket has to be in /run/p11-kit directory", slot.Name)
```

Comment:
nitpick: on this and in the above message, the "a" makes one think that there's more than a socket path in the slot, whereas -- looking at the code -- there's just one.

Also, in the message about the path not being clean, you could print the path itself.